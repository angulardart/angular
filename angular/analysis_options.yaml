# This is a copy of ../analysis_options.yaml, with implicit-casts: false, and an "exclude: ...".

analyzer:
  strong-mode:
    implicit-casts: false
  exclude:
    - "tools/analyzer_plugin/bin/plugin.dart"
  errors:
    unused_element: error
    unused_import: error
    unused_local_variable: error
    dead_code: error

    # Allow importing .template.dart files without an analyzer error.
    uri_has_not_been_generated: ignore

    # The noise of deprecations isn't worth the effort.
    deprecated_member_use_from_same_package: ignore
linter:
  rules:
    - avoid_empty_else
    - avoid_init_to_null
    - avoid_null_checks_in_equality_operators
    - await_only_futures
    - camel_case_types
    - cancel_subscriptions
    - control_flow_in_finally
    - directives_ordering
    - empty_catches
    - empty_constructor_bodies
    - empty_statements
    - hash_and_equals
    - iterable_contains_unrelated_type
    - library_names
    - library_prefixes
    - list_remove_unrelated_type
    - only_throw_errors
    - overridden_fields
    - package_api_docs
    - package_names
    - package_prefixed_library_names
    - prefer_final_fields
    - prefer_is_not_empty
    - slash_for_doc_comments
    - test_types_in_equals
    - throw_in_finally
    - type_init_formals
    - unnecessary_brace_in_string_interps
    - unrelated_type_equality_checks
    - valid_regexps
