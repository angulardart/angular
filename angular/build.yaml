targets:
  $default:
    builders:
      angular:
        generate_for:
          exclude:
            - "lib/src/build.dart"
            - "lib/src/compiler/**"
            - "lib/src/source_gen/**"

builders:
  angular:
    import: "package:angular/src/build.dart"
    builder_factories:
      - templatePlaceholder
      - templateCompiler
      - stylesheetCompiler
    auto_apply: dependents
    applies_builders:
      - "angular|placeholder_cleanup"
      - "angular|component_source_cleanup"
    # See https://github.com/dart-lang/angular/issues/988.
    is_optional: true
    required_inputs:
      - ".css"
    build_extensions:
      .css:
        - ".css.dart"
        - ".css.shim.dart"
      .dart:
        - ".template.dart"
post_process_builders:
  placeholder_cleanup:
    import: "package:angular/src/build.dart"
    builder_factory: "placeholderCleanup"
  component_source_cleanup:
    import: "package:angular/src/build.dart"
    builder_factory: "componentSourceCleanup"
    defaults:
      generate_for: ["lib/**"]
      release_options:
        enabled: true
