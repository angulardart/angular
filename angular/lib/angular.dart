/// The primary library for the [AngularDart web framework][AngularDart].
///
/// Import this library as follows:
///
/// ```
/// import 'package:angular/angular.dart';
/// ```
///
/// For help using this library, see the AngularDart documentation:
///
/// * [AngularDart guide][]
/// * [AngularDart cheat sheet][cheatsheet]
///
/// [AngularDart]: https://webdev.dartlang.org/angular
/// [AngularDart guide]: https://webdev.dartlang.org/angular/guide
/// [cheatsheet]: https://webdev.dartlang.org/angular/cheatsheet

library angular;

export 'core.dart';
export 'src/common/common_directives.dart';
export 'src/common/directives.dart';
export 'src/common/pipes.dart';
export 'src/core/angular_entrypoint.dart' show AngularEntrypoint;
export 'src/core/application_tokens.dart' hide APP_ID_RANDOM_PROVIDER;
export 'src/platform/bootstrap.dart';
export 'src/platform/dom/dom_tokens.dart';
export 'src/platform/dom/events/event_manager.dart' show EventManagerPlugin;
