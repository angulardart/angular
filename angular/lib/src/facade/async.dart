import 'dart:async';

/// **DEPRECATED**: Use [StreamController] and [Stream] instead.
///
/// **BEFORE**:
/// ```
/// @Output()
/// final touch = new EventEmitter();
/// ```
///
/// **AFTER**:
/// ```
/// final _onTouch = new StreamController.broadcast();
///
/// @Output()
/// Stream get touch => _onTouch.stream;
/// ```
///
/// This avoids leaking details about how your [Stream] is implemented.
@Deprecated('Only expose a Stream for your component @Output')
class EventEmitter<T> extends Stream<T> {
  StreamController<T> _controller;

  /// Creates an instance of [EventEmitter], which depending on [isAsync],
  /// delivers events synchronously or asynchronously.
  EventEmitter([bool isAsync = true]) {
    _controller = new StreamController<T>.broadcast(sync: !isAsync);
  }

  StreamSubscription<T> listen(void onData(T event),
      {Function onError, void onDone(), bool cancelOnError}) {
    return _controller.stream.listen(onData,
        onError: onError, onDone: onDone, cancelOnError: cancelOnError);
  }

  void add(T value) {
    _controller.add(value);
  }

  void emit(T value) {
    _controller.add(value);
  }

  void addError(error) {
    _controller.addError(error);
  }

  void close() {
    _controller.close();
  }
}
