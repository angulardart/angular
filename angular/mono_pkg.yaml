dart:
- stable
- dev

stages:
- analyze:
  - dartanalyzer: --fatal-warnings .
