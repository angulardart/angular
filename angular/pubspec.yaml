name: angular
version: 5.3.0
author: Dart Team <web@dartlang.org>
description: Fast and productive web framework
homepage: https://webdev.dartlang.org/angular
documentation: https://webdev.dartlang.org/angular/api
environment:
  sdk: ">=2.2.1-dev.3.0 <3.0.0"

dependencies:
  # Runtime (or runtime *and* the compiler).
  collection: ^1.14.5
  intl: '>=0.13.0 <0.16.0'
  js: ^0.6.0
  logging: '>=0.9.0 <0.12.0'
  meta: ^1.1.4
  stack_trace: ^1.9.1

  # Compiler. Eventually we want to move these to angular_compiler.
  analyzer: ^0.36.2

  # It is extremely important these two dependencies stay pinned to current.
  angular_ast: 0.5.9
  angular_compiler: 0.4.3
  ##########################################################################

  build: '>=0.12.7 <2.0.0'
  build_config: '>=0.2.6 <0.5.0' # Runtime dependency with no import
  code_builder: '^3.0.1'
  csslib: '>=0.14.5 <0.16.0'
  path: ^1.6.1
  source_gen: ">=0.9.1 <0.10.0"
  source_span: ^1.4.0

# === vvv REMOVE WHEN PUBLISHING vvv ===
dependency_overrides:
  angular_ast:
      path: ../angular_ast
  angular_compiler:
      path: ../angular_compiler
# === ^^^ REMOVE WHEN PUBLISHING ^^^ ===
