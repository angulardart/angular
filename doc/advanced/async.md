# Async Updates


In most apps and components, it is not necessary to precisely know how
asynchronous events, as propagated through events, callbacks,
[`Future`s][futures], or [`Stream`s][streams] interact with AngularDart, as they
are handled automatically.

[futures]: https://www.dartlang.org/tutorials/language/futures
[streams]: https://www.dartlang.org/tutorials/language/streams

WARNING: The advice and recommendations in this article pertain to users of
_default_ change detection. Components that use `ChangeDetectionStrategy.OnPush`
should inject `ChangeDetectorRef` and call `markForCheck()` as a signal whenever
necessary.

In _some_ cases, you may want to inject [`NgZone`][ng-zone] (as a constructor
parameter) to control how AngularDart uses [zones][zones] to observe your
application. This can help with bugs such as _I updated my state, but my DOM
does not update until X happens_.

[ng-zone]: https://webdev.dartlang.org/api/angular/angular/NgZone-class
[zones]: https://www.dartlang.org/articles/libraries/zones

## Making Angular aware of async callbacks

When using [`Future`][futures], [`Stream`][streams], or event bindings in
templates, Angular is automatically aware when an event completes and will use
that as a signal to rerun change detection. In some cases, though, you may have
code that uses callbacks or JS-interop that is not aware of zones.

**NOTE**: Angular is not aware of DOM events, but reenters the Angular zone
automatically when handling them in templates. DOM event listeners registered in
Dart code must manually reenter the zone to trigger change detection as
demonstrated below.

Imagine you have the following code, perhaps in another shared library:

```dart
class AsyncNotifier {
  final _listeners = <void Function()>[];

  void registerListener(void Function() listener) {
    _listeners.add(listener);
  }

  void notifyListeners() {
    for (final listener in _listeners) {
      listener();
    }
  }
}
```

When using it inside an AngularDart component or service, this code does not
respect [Zone][zones] semantics, and AngularDart might not be aware of when
`notifyListeners` was invoked:

```dart
import 'package:angular/angular.dart';

@Component(
  selector: 'notify-listener',
  template: 'I was notified {{count}} times',
)
class NotifyListener {
  var count = 0;

  @Input()
  set notifier(AsyncNotifier notifier) {
    notifier.registerListener(() {
      count++;
      print('New count is $count');
    });
  }
}
```

For example, you might see `New count is 1` in your development console but in
the DOM observe `I was notified 0 times`. In cases like this, you want to ensure
your callback executes within the Angular observed zone, and you can use
`<NgZone>.run`:

```dart
import 'package:angular/angular.dart';

@Component(
  selector: 'notify-listener',
  template: 'I was notified {{count}} times',
)
class NotifyListener {
  final NgZone _ngZone;

  NotifyListener(this._ngZone);

  var count = 0;

  @Input()
  set notifier(AsyncNotifier notifier) {
    notifier.registerListener(() {
      _ngZone.run(() {
        count++;
        print('New count is $count');
      });
    });
  }
}
```

You should now see the DOM and the development console updated. This pattern is
also useful for JS-interop or any other library or API that involves registering
callbacks/closures that might not be aware of Angular.

**NOTE**: If you are using `ChangeDetectionStrategy.OnPush`, don't forget to
inject and use `<ChangeDetectorRef>.markForCheck` when reentering the `NgZone`:

```dart
import 'package:angular/angular.dart';

@Component(
  selector: 'notify-listener',
  template: 'I was notified {{count}} times',
  changeDetection: ChangeDetectionStrategy.OnPush,
)
class NotifyListener {
  final ChangeDetectorRef _changeDetector;
  final NgZone _ngZone;

  NotifyListener(this._changeDetector, this._ngZone);

  var count = 0;

  @Input()
  set notifier(AsyncNotifier notifier) {
    notifier.registerListener(() {
      _ngZone.run(() {
        count++;
        print('New count is $count');
        _changeDetector.markForCheck();
      });
    });
  }
}
```

## Ignoring or filtering async callbacks

In some cases, events might be fired at a high frequency that would negatively
effect performance if Angular reacted to each of them, and you want to
completely (or selectively) ignore events:

```dart
import 'dart:html';

import 'package:angular/angular.dart';

@Component(
  selector: 'custom-painter',
  template: '...',
)
class CustomPainter implements OnInit {
  final HtmlElement _element;

  CustomPainter(this._element);

  @override
  void ngOnInit() {
    _element.onMouseMove.listen((_) {
      // Do something
    });
  }
}
```

While this code works, _every time_ `.onMouseMove` fires AngularDart will
trigger an update (unless you are using `ChangeDetectionStrategy.OnPush`), so
you may use `<NgZone>.runOutsideAngular` to ignore these events:

```dart
import 'dart:html';

import 'package:angular/angular.dart';

@Component(
  selector: 'custom-painter',
  template: '...',
)
class CustomPainter implements OnInit {
  final HtmlElement _element;
  final NgZone _ngZone;

  CustomPainter(this._element, this._ngZone);

  @override
  void ngOnInit() {
    _ngZone.runOutsideAngular(() {
      _element.onMouseMove.listen((_) {
        // Do something
      });
    });
  }
}
```

You can combine with `<NgZone>.run` to selectively re-enter Angular:

```dart
@Component(
  selector: 'custom-painter',
  template: '...',
)
class CustomPainter implements OnInit {
  final HtmlElement _element;
  final NgZone _ngZone;

  CustomPainter(this._element, this._ngZone);

  @override
  void ngOnInit() {
    _ngZone.runOutsideAngular(() {
      _element.onMouseMove.listen((event) {
        if (event.clientX > 500) {
          _ngZone.run(() {
            // Do something
          });
        }
      });
    });
  }
}
```

## Waiting for state to propagate

You may run into scenarios where you want to wait for change detection to be
executed before continuing, and can use `<NgZone>.runAfterChangesObserved`:

```dart
import 'dart:html';

import 'package:angular/angular.dart';

@Component(
  selector: 'example-state',
  template: r'''
    <div [style.height]="height" #someDiv></div>
  ''',
)
class Example {
  final NgZone _ngZone;

  Example(this._ngZone);

  int height = 100;

  @ViewChild('someDiv')
  set someDiv(HtmlElement someDiv) {
    if (someCondition) {
      height = 200;
      _ngZone.runAfterChangesObserved(() {
        checkDimensionsOf(someDiv);
      });
    }
  }
}
```

**NOTE**: It might be tempting to use `scheduleMicrotask`, `await null`,
`Timer.run`, or other more nefarious API surfaces like
`<ChangeDetectorRef>.detectChanges()` to accomplish this, but resist the
temptation as these methods have negative implications for performance and
stability. See also [`Prefer NgZone.runAfterChangesObserved`][prefer] for
details.

[prefer]: ../effective/change-detection.md?#prefer-ngzonerunafterchangesobserved-to-timers-or-microtasks
