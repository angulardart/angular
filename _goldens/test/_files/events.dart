import 'dart:async';

import 'package:angular/angular.dart';

@Component(
  selector: 'uses-native-events',
  template: r'''
    <button (click)="onClick()"></button>
  ''',
)
class UsesNativeEvents {
  @HostListener('focus')
  void onFocus() {}

  void onClick() {}
}

@Component(
  selector: 'uses-angular-events',
  directives: [
    HasAngularEvents,
  ],
  template: r'''
    <has-angular-events (foo)="onFoo()"></has-angular-events>
  ''',
)
class UsesAngularEvents {
  void onFoo() {}
}

@Component(
  selector: 'has-angular-events',
  template: '',
)
class HasAngularEvents {
  @Output()
  Stream<Null> get foo => const Stream.empty();
}

@Component(
  selector: 'material-button-like',
  template: '',
)
class HasManyNativeHostEvents {
  @HostListener('mousedown')
  void onMouseDown(Object e) {}

  @HostListener('mouseup')
  void onMouseUp(Object e) {}

  @HostListener('click')
  void onClick(Object e) {}

  @HostListener('keypress')
  void onKeyPress(Object e) {}

  @HostListener('focus')
  void onFocus(Object e) {}

  @HostListener('blur')
  void onBlur(Object e) {}
}
