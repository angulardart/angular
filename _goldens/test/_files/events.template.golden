// **************************************************************************
// Generator: AngularDart Compiler
// **************************************************************************

import 'events.dart';
export 'events.dart';
import 'package:angular/src/di/reflector.dart' as _ngRef;
import 'package:angular/angular.template.dart' as _ref0;
import 'package:angular/src/core/linker/views/component_view.dart' as import0;
import 'events.dart' as import1;
import 'package:angular/src/core/linker/style_encapsulation.dart' as import2;
import 'package:angular/src/core/linker/views/view.dart' as import3;
import 'package:angular/src/core/change_detection/change_detection.dart';
import 'dart:html' as import5;
import 'package:angular/src/runtime.dart' as import6;
import 'package:angular/src/runtime/dom_helpers.dart' as import7;
import 'package:angular/angular.dart';
import 'package:angular/src/core/linker/views/host_view.dart' as import9;

final List<dynamic> styles$UsesNativeEvents = const [];

// View for component UsesNativeEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class ViewUsesNativeEvents0 extends import0.ComponentView<import1.UsesNativeEvents> {
  static import2.ComponentStyles _componentStyles;
  ViewUsesNativeEvents0(import3.View parentView, int parentIndex) : super(parentView, parentIndex, ChangeDetectionStrategy.CheckAlways) {
    this.initComponentStyles();
    this.rootElement = import5.document.createElement('uses-native-events');
  }
  static String get _debugComponentUrl {
    return (import6.isDevMode ? 'asset:_goldens/test/_files/events.dart' : null);
  }

  @override
  void build() {
    final _ctx = this.ctx;
    final import5.HtmlElement parentRenderNode = this.initViewRoot();
    final doc = import5.document;
    final _el_0 = import7.appendElement(doc, parentRenderNode, 'button');
    _el_0.addEventListener('click', this.eventHandler0(_ctx.onClick));
    parentRenderNode.addEventListener('focus', this.eventHandler0(_ctx.onFocus));
  }

  static void _debugClearComponentStyles() {
    _componentStyles = null;
  }

  void initComponentStyles() {
    var styles = _componentStyles;
    if (identical(styles, null)) {
      _componentStyles = (styles = import2.ComponentStyles.unscoped(styles$UsesNativeEvents, _debugComponentUrl));
      if (import6.isDevMode) {
        import2.ComponentStyles.debugOnClear(_debugClearComponentStyles);
      }
    }
    this.componentStyles = styles;
  }
}

const ComponentFactory<import1.UsesNativeEvents> _UsesNativeEventsNgFactory = const ComponentFactory('uses-native-events', viewFactory_UsesNativeEventsHost0);
ComponentFactory<import1.UsesNativeEvents> get UsesNativeEventsNgFactory {
  return _UsesNativeEventsNgFactory;
}

final List<dynamic> styles$UsesNativeEventsHost = const [];

// Host for component UsesNativeEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class _ViewUsesNativeEventsHost0 extends import9.HostView<import1.UsesNativeEvents> {
  @override
  void build() {
    // Child component UsesNativeEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
    this.componentView = ViewUsesNativeEvents0(this, 0);
    final _el_0 = this.componentView.rootElement;
    this.component = import1.UsesNativeEvents();
    this.initRootNode(_el_0);
  }
}

import9.HostView<import1.UsesNativeEvents> viewFactory_UsesNativeEventsHost0() {
  return _ViewUsesNativeEventsHost0();
}

final List<dynamic> styles$UsesAngularEvents = const [];

// View for component UsesAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class ViewUsesAngularEvents0 extends import0.ComponentView<import1.UsesAngularEvents> {
  ViewHasAngularEvents0 _compView_0;
  import1.HasAngularEvents _HasAngularEvents_0_5;
  static import2.ComponentStyles _componentStyles;
  ViewUsesAngularEvents0(import3.View parentView, int parentIndex) : super(parentView, parentIndex, ChangeDetectionStrategy.CheckAlways) {
    this.initComponentStyles();
    this.rootElement = import5.document.createElement('uses-angular-events');
  }
  static String get _debugComponentUrl {
    return (import6.isDevMode ? 'asset:_goldens/test/_files/events.dart' : null);
  }

  @override
  void build() {
    final _ctx = this.ctx;
    final import5.HtmlElement parentRenderNode = this.initViewRoot();
    // Child component HasAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
    this._compView_0 = ViewHasAngularEvents0(this, 0);
    final _el_0 = this._compView_0.rootElement;
    parentRenderNode.append(_el_0);
    this._HasAngularEvents_0_5 = import1.HasAngularEvents();
    this._compView_0.create(this._HasAngularEvents_0_5);
    final subscription_0 = this._HasAngularEvents_0_5.foo.listen(this.eventHandler0(_ctx.onFoo));
    this.initSubscriptions([subscription_0]);
  }

  @override
  void detectChangesInternal() {
    this._compView_0.detectChanges();
  }

  @override
  void destroyInternal() {
    this._compView_0.destroyInternalState();
  }

  static void _debugClearComponentStyles() {
    _componentStyles = null;
  }

  void initComponentStyles() {
    var styles = _componentStyles;
    if (identical(styles, null)) {
      _componentStyles = (styles = import2.ComponentStyles.unscoped(styles$UsesAngularEvents, _debugComponentUrl));
      if (import6.isDevMode) {
        import2.ComponentStyles.debugOnClear(_debugClearComponentStyles);
      }
    }
    this.componentStyles = styles;
  }
}

const ComponentFactory<import1.UsesAngularEvents> _UsesAngularEventsNgFactory = const ComponentFactory('uses-angular-events', viewFactory_UsesAngularEventsHost0);
ComponentFactory<import1.UsesAngularEvents> get UsesAngularEventsNgFactory {
  return _UsesAngularEventsNgFactory;
}

final List<dynamic> styles$UsesAngularEventsHost = const [];

// Host for component UsesAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class _ViewUsesAngularEventsHost0 extends import9.HostView<import1.UsesAngularEvents> {
  @override
  void build() {
    // Child component UsesAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
    this.componentView = ViewUsesAngularEvents0(this, 0);
    final _el_0 = this.componentView.rootElement;
    this.component = import1.UsesAngularEvents();
    this.initRootNode(_el_0);
  }
}

import9.HostView<import1.UsesAngularEvents> viewFactory_UsesAngularEventsHost0() {
  return _ViewUsesAngularEventsHost0();
}

final List<dynamic> styles$HasAngularEvents = const [];

// View for component HasAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class ViewHasAngularEvents0 extends import0.ComponentView<import1.HasAngularEvents> {
  static import2.ComponentStyles _componentStyles;
  ViewHasAngularEvents0(import3.View parentView, int parentIndex) : super(parentView, parentIndex, ChangeDetectionStrategy.CheckAlways) {
    this.initComponentStyles();
    this.rootElement = import5.document.createElement('has-angular-events');
  }
  static String get _debugComponentUrl {
    return (import6.isDevMode ? 'asset:_goldens/test/_files/events.dart' : null);
  }

  @override
  void build() {
    final import5.HtmlElement parentRenderNode = this.initViewRoot();
  }

  static void _debugClearComponentStyles() {
    _componentStyles = null;
  }

  void initComponentStyles() {
    var styles = _componentStyles;
    if (identical(styles, null)) {
      _componentStyles = (styles = import2.ComponentStyles.unscoped(styles$HasAngularEvents, _debugComponentUrl));
      if (import6.isDevMode) {
        import2.ComponentStyles.debugOnClear(_debugClearComponentStyles);
      }
    }
    this.componentStyles = styles;
  }
}

const ComponentFactory<import1.HasAngularEvents> _HasAngularEventsNgFactory = const ComponentFactory('has-angular-events', viewFactory_HasAngularEventsHost0);
ComponentFactory<import1.HasAngularEvents> get HasAngularEventsNgFactory {
  return _HasAngularEventsNgFactory;
}

final List<dynamic> styles$HasAngularEventsHost = const [];

// Host for component HasAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class _ViewHasAngularEventsHost0 extends import9.HostView<import1.HasAngularEvents> {
  @override
  void build() {
    // Child component HasAngularEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
    this.componentView = ViewHasAngularEvents0(this, 0);
    final _el_0 = this.componentView.rootElement;
    this.component = import1.HasAngularEvents();
    this.initRootNode(_el_0);
  }
}

import9.HostView<import1.HasAngularEvents> viewFactory_HasAngularEventsHost0() {
  return _ViewHasAngularEventsHost0();
}

final List<dynamic> styles$HasManyNativeHostEvents = const [];

// View for component HasManyNativeHostEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class ViewHasManyNativeHostEvents0 extends import0.ComponentView<import1.HasManyNativeHostEvents> {
  static import2.ComponentStyles _componentStyles;
  ViewHasManyNativeHostEvents0(import3.View parentView, int parentIndex) : super(parentView, parentIndex, ChangeDetectionStrategy.CheckAlways) {
    this.initComponentStyles();
    this.rootElement = import5.document.createElement('material-button-like');
  }
  static String get _debugComponentUrl {
    return (import6.isDevMode ? 'asset:_goldens/test/_files/events.dart' : null);
  }

  @override
  void build() {
    final _ctx = this.ctx;
    final import5.HtmlElement parentRenderNode = this.initViewRoot();
    parentRenderNode.addEventListener('mousedown', this.eventHandler1(_ctx.onMouseDown));
    parentRenderNode.addEventListener('mouseup', this.eventHandler1(_ctx.onMouseUp));
    parentRenderNode.addEventListener('click', this.eventHandler1(_ctx.onClick));
    parentRenderNode.addEventListener('keypress', this.eventHandler1(_ctx.onKeyPress));
    parentRenderNode.addEventListener('focus', this.eventHandler1(_ctx.onFocus));
    parentRenderNode.addEventListener('blur', this.eventHandler1(_ctx.onBlur));
  }

  static void _debugClearComponentStyles() {
    _componentStyles = null;
  }

  void initComponentStyles() {
    var styles = _componentStyles;
    if (identical(styles, null)) {
      _componentStyles = (styles = import2.ComponentStyles.unscoped(styles$HasManyNativeHostEvents, _debugComponentUrl));
      if (import6.isDevMode) {
        import2.ComponentStyles.debugOnClear(_debugClearComponentStyles);
      }
    }
    this.componentStyles = styles;
  }
}

const ComponentFactory<import1.HasManyNativeHostEvents> _HasManyNativeHostEventsNgFactory = const ComponentFactory('material-button-like', viewFactory_HasManyNativeHostEventsHost0);
ComponentFactory<import1.HasManyNativeHostEvents> get HasManyNativeHostEventsNgFactory {
  return _HasManyNativeHostEventsNgFactory;
}

final List<dynamic> styles$HasManyNativeHostEventsHost = const [];

// Host for component HasManyNativeHostEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
class _ViewHasManyNativeHostEventsHost0 extends import9.HostView<import1.HasManyNativeHostEvents> {
  @override
  void build() {
    // Child component HasManyNativeHostEvents in asset:_goldens/test/_files/events.dart (changeDetection: Default)
    this.componentView = ViewHasManyNativeHostEvents0(this, 0);
    final _el_0 = this.componentView.rootElement;
    this.component = import1.HasManyNativeHostEvents();
    this.initRootNode(_el_0);
  }
}

import9.HostView<import1.HasManyNativeHostEvents> viewFactory_HasManyNativeHostEventsHost0() {
  return _ViewHasManyNativeHostEventsHost0();
}

var _visited = false;
void initReflector() {
  if (_visited) {
    return;
  }
  _visited = true;

  _ngRef.registerComponent(UsesNativeEvents, UsesNativeEventsNgFactory);
  _ngRef.registerComponent(UsesAngularEvents, UsesAngularEventsNgFactory);
  _ngRef.registerComponent(HasAngularEvents, HasAngularEventsNgFactory);
  _ngRef.registerComponent(HasManyNativeHostEvents, HasManyNativeHostEventsNgFactory);
  _ref0.initReflector();
}
